Add Audio Interruption Handling for Camera Recording

morepriyam opened the pull request:
## Summary

Implements comprehensive audio interruption handling for video recording using `expo-audio` to manage audio sessions and `AppState` monitoring to handle background transitions gracefully. This ensures recordings stop automatically when the app is backgrounded and all UI states reset to a clean initial state when the app becomes active again.

## Problem

Previously, the app had no handling for:
- ❌ App going to background during recording (recording would continue)
- ❌ Phone calls interrupting recording
- ❌ Other apps taking audio focus
- ❌ UI animations/states not resetting after interruptions
- ❌ No audio session management for better interruption resilience

## Solution

### 1. Audio Session Configuration (`expo-audio`)
- Added `expo-audio` package and plugin configuration
- Configure audio session before each recording with `setAudioModeAsync()`
- Use `interruptionMode: "doNotMix"` to request exclusive audio focus
- Ensures other apps pause their audio when recording starts

### 2. Background/Foreground Handling
- Added `AppState` listener to detect when app goes to background/inactive
- Automatically stops recording when app is backgrounded
- Manages audio subsystem (enable/disable) based on app state
- Resets all animations and UI states when app becomes active again

### 3. Error Handling Improvements
- Added camera mount error handling with retry option
- Enhanced recording promise error handling for interruptions
- Added comprehensive debug logging with emoji markers for easy identification

### 4. State Reset on App Activation
- Added `reset()` method to `RecordButton` component (exposed via ref)
- Resets all animation values to initial state
- Resets zoom, touch states, and recording states
- Ensures clean UI state when returning from background

## Changes

### New Dependencies
- `expo-audio` (~1.1.1) - Audio session management

### Files Modified

#### `app.json`
- Added `expo-audio` plugin configuration with microphone permission

#### `components/RecordButton.tsx`
- Added `expo-audio` import and `setAudioModeAsync()` call
- Configure audio session before starting recording
- Converted to `forwardRef` to expose `reset()` method
- Added `reset()` function to reset all animations and states
- Enhanced error handling for interruptions
- Added debug logging

#### `app/(camera)/shorts.tsx`
- Added `AppState` monitoring during recording
- Stop recording when app goes to background/inactive
- Reset all states when app becomes active again
- Added camera mount error handling
- Added `onCameraReady` callback
- Added `recordButtonRef` to access reset functionality

## Technical Details

### Audio Session Configuration
```typescript
await setAudioModeAsync({
  allowsRecording: true,
  playsInSilentMode: true,
  interruptionMode: "doNotMix", // Exclusive audio focus
  shouldPlayInBackground: false,
});
```

### AppState Monitoring
```typescript
// Stops recording on background
if (nextAppState === "background" || nextAppState === "inactive") {
  cameraRef.current?.stopRecording();
  await setIsAudioActiveAsync(false);
}

// Resets state on active
if (nextAppState === "active") {
  await setIsAudioActiveAsync(true);
  recordButtonRef.current?.reset();
  // Reset zoom, touch states, etc.
}
```

## Testing

### Test Scenarios
1. ✅ **Background Interruption**: Start recording → Press home button → Recording stops automatically
2. ✅ **Audio Session**: Console shows "Audio session configured successfully" when recording starts
3. ✅ **State Reset**: Return to app → All animations and states reset to initial values
4. ✅ **Phone Calls**: Recording handles phone call interruptions gracefully
5. ✅ **Error Handling**: Camera mount errors show retry option

### Console Logs to Verify
- `[RecordButton] ✅ Audio session configured successfully`
- `[ShortsScreen] 📱 App state changed to: inactive`
- `[ShortsScreen] 🛑 Stopping recording due to background/inactive state`
- `[RecordButton] 🔄 Reset all animations and states`
- `[ShortsScreen] 🔄 Reset all animations and states`

## Benefits

- ✅ **Better UX**: Recordings stop automatically when app is backgrounded
- ✅ **Clean State**: All UI resets to initial state when returning from background
- ✅ **Audio Focus**: Exclusive audio focus prevents conflicts with other apps
- ✅ **Error Recovery**: Better error handling and recovery options
- ✅ **Debugging**: Comprehensive logging for troubleshooting

## Breaking Changes

None - This is a backward-compatible enhancement.

## Screenshots/Logs

### Before
- Recording continues when app goes to background
- No audio session management
- UI states can be inconsistent after interruptions

### After
- Recording stops automatically on background
- Audio session properly configured
- Clean state reset on app activation

## Related Issues

Fixes issues with:
- Recording continuing in background
- Audio conflicts with other apps
- Inconsistent UI state after interruptions

## Checklist

- [x] Code follows project style guidelines
- [x] Self-review completed
- [x] Comments added for complex logic
- [x] Documentation updated (if applicable)
- [x] No new warnings generated
- [x] Tests pass (manual testing completed)
- [x] Build succeeds

## Notes

- `expo-audio` is used primarily for audio session configuration, not for recording (expo-camera handles recording)
- The `doNotMix` interruption mode ensures exclusive audio focus, which is ideal for video recording
- State reset ensures consistent UI behavior when users return to the app

morepriyam committed:
Enhance Audio Recording Capabilities and Permissions

- Added expo-audio dependency to manage audio recording and permissions.
- Updated app.json to include microphone permission description for audio recording.
- Modified AndroidManifest.xml to request MODIFY_AUDIO_SETTINGS permission.
- Implemented audio session configuration in RecordButton for better interruption handling.
- Enhanced ShortsScreen to manage audio state during app lifecycle changes.
- Added error handling for camera mount issues and improved user feedback.

This update improves the audio recording experience and ensures proper handling of audio interruptions during recording sessions.

adithya1012 commented:
Hello @morepriyam 👋

I tried testing this branch locally on **Android**, but it doesn’t seem to be working as expected.

### Observed behavior

The app builds successfully, but video recording fails on Android. Below are the logs I’m seeing:

```
Android Bundled 114ms node_modules/expo-router/entry.js (1 module)
LOG  🔗 Deeplink params: {}
LOG  [ShortsScreen] ✅ Camera ready
LOG  [RecordButton] ✅ Audio session configured successfully
LOG  [RecordButton] 🔄 Reset all animations and states
LOG  [ShortsScreen] 🔄 Reset all animations and states
LOG  [RecordButton] 🔄 Reset all animations and states
LOG  [ShortsScreen] 🔄 Reset all animations and states
ERROR  [RecordButton] ❌ Recording failed (unexpected): 
[Error: Video recording failed: Video recording Failed: Unknown error]
LOG  [RecordButton] 🔄 Reset all animations and states
LOG  [ShortsScreen] 🔄 Reset all animations and states
```

### Screenshot

<img width="193" height="426" alt="image" src="https://github.com/user-attachments/assets/a24cde43-0e75-4b3f-9977-200723c41822" />

### Steps to reproduce (Android)

1. `git checkout 257-audio-issues-in-camera`
2. `npm install`
3. `npx expo run:android`

Please let me know if I’m missing any setup steps on Android.
I’m also looking into this from my side to see if I can identify or fix the issue.

Thanks!

adithya1012 reviewed (changes requested):
Hello @morepriyam 👋

I tried testing this branch locally on **Android**, but it doesn’t seem to be working as expected.

### Observed behavior

The app builds successfully, but video recording fails on Android. Below are the logs I’m seeing:

```
Android Bundled 114ms node_modules/expo-router/entry.js (1 module)
LOG  🔗 Deeplink params: {}
LOG  [ShortsScreen] ✅ Camera ready
LOG  [RecordButton] ✅ Audio session configured successfully
LOG  [RecordButton] 🔄 Reset all animations and states
LOG  [ShortsScreen] 🔄 Reset all animations and states
LOG  [RecordButton] 🔄 Reset all animations and states
LOG  [ShortsScreen] 🔄 Reset all animations and states
ERROR  [RecordButton] ❌ Recording failed (unexpected): 
[Error: Video recording failed: Video recording Failed: Unknown error]
LOG  [RecordButton] 🔄 Reset all animations and states
LOG  [ShortsScreen] 🔄 Reset all animations and states
```

### Screenshot

<img width="193" height="426" alt="image" src="https://github.com/user-attachments/assets/a24cde43-0e75-4b3f-9977-200723c41822" />

### Steps to reproduce (Android)

1. `git checkout 257-audio-issues-in-camera`
2. `npm install`
3. `npx expo run:android`

Please let me know if I’m missing any setup steps on Android.
I’m also looking into this from my side to see if I can identify or fix the issue.

Thanks!

adithya1012 committed:
fix(android): prevent AppState from rapidly switching during recording

On Android, React Native's AppState incorrectly triggers background/active
transitions when any UI overlay appears
causing recording to fail with "Video recording Failed: Unknown error".

Changes:

- Track previous app state to detect genuine background transitions
- Add 500ms threshold to filter rapid false state changes on Android
- Use 'focus' event on Android instead of 'change' event for more reliable detection
- Only reset recording state on genuine background→active transitions

horner commented:
The first comment on the PR should be a summary of the PR at the latest moment with a screenshot(s) and video short.

horner commented:
I wonder if we shouldn't try to block incoming phone calls if opposed to blocking the recording

morepriyam committed:
Merge branch 'main' into 257-audio-issues-in-camera
